website/blog/2019-07-15-complete-table-scan.md

```diff
@@ -0,0 +1,174 @@
+```
+| Version              | Wall time (seconds) | CPU time (seconds) | Baseline CPU / Aria CPU |
+| -------------------- | ------------------- | ------------------ | ----------------------- |
+| Aria                 | 5                   | 52                 | 1.0                     |
+| Aria w/no reordering | 5                   | 57                 | 1.09                    |
+| Baseline             | 7                   | 96                 | 1.84                    |
+
+Here we have a conjunction of low selectivity filters, each true approximately 9/10 of the time. We are running Aria with and without filter reordering. We see about 10% gain from reordering even in a situation where the difference between different filter orders is small. We discussed the basic principle of reordering in [Everything You Always Wanted To Do in Table Scan](/blog/2019/06/29/everything-you-always-wanted-to-do-in-a-table-scan). Measure the time per dropped row and put the fastest first.
+
+```sql
+SELECT COUNT(*), MAX(partkey)
+FROM hive.tpch.lineitem
+WHERE comment LIKE '%fur%' AND partkey + 1 < 19000000 AND suppkey + 1 < 100000;
+```
+| Version              | Wall time (seconds) | CPU time (seconds) | Baseline CPU / Aria CPU |
+| -------------------- | ------------------- | ------------------ | ----------------------- |
+| Aria                 | 7                   | 98.7               | 1.0                     |
+| Aria w/no reordering | 24                  | 339                | 3.43                    |
+| Baseline             | 25                  | 361                | 3.65                    |
+
+Here we have a case where adaptivity makes a large difference with expensive filter expressions. The key to the puzzle is that the conjunct on `partkey` is cheap and selects 1/10. The like is 1/5 and expensive. The `partkey` is cheap and 19/20. Putting `suppkey` first wins the race. This happens without any reliance on statistics or pre-execution optimization and will adapt at run time if the data properties change.
+
+## More flexible filters
+
+```sql
+SELECT COUNT(*), SUM(extendedprice)
+FROM lineitem
+WHERE shipmode LIKE '%AIR%' AND shipinstruct LIKE '%PERSON';
+```
+| Version  | Wall time (seconds) | CPU time (seconds) | Baseline CPU / Aria CPU |
+| -------- | ------------------- | ------------------ | ----------------------- |
+| Aria     | 4                   | 44.2               | 1.0                     |
+| Baseline | 21                  | 271                | 6.13                    |
+
+The filtered columns are of low cardinality and are encoded as dictionaries. This is an example of evaluating an expensive predicate on only distinct values. Baseline Presto misses the opportunity because all filters are generated into a monolithic code block. Aria generates filter expressions for each distinct set of required columns. In this case the filters are independent and reorderable.
```

Comment:
@oerling I didn't realize that complex filters also run on dictionaries. This is super cool. Do you have a pointer for me to check out how this is done?